### ISSUE TITLE
bug: Exception is thrown when creating Kinesis stream through Cloudformation template

### ISSUE THREAD

**Issue by ayeganyan** [localstack/localstack#8404]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I see the following exception in the logs when I create Kinesis stream through Cloudformation template:
```
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : [io-compute-0] WARN  2023-05-31 12:57:49,199 k.m.cache.Cache x-amzn-RequestId=befeeb12-ffb2-11ed-abd3-31560c049676, action=DescribeStream, contextId=befeeb11-ffb2-11ed-abd3-31560c049676, x-amz-id-2=uXqA7jQMjz7bgPU94nrmzSJ+r7d6nl1nKYCJeZyJrVeqsm56WG9d4I4WK1aCKg3im5reKFG63W+BIKB6ksySRgp0CmNBIMXT, contentType=application/x-amz-json-1.1, streamName=kinesis-test - Describing the stream was unuccessful
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : kinesis.mock.ResourceNotFoundException: Stream arn arn:aws:kinesis:us-east-1:000000000000:stream/dx-ref-cfn-kinesis1 not found
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at kinesis.mock.validations.CommonValidations$.$anonfun$findStream$1(CommonValidations.scala:72)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at scala.Option.toRight(Option.scala:593)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at kinesis.mock.validations.CommonValidations$.findStream(CommonValidations.scala:72)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at kinesis.mock.api.DescribeStreamRequest.$anonfun$describeStream$3(DescribeStreamRequest.scala:32)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at scala.util.Either.flatMap(Either.scala:352)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at kinesis.mock.api.DescribeStreamRequest.$anonfun$describeStream$2(DescribeStreamRequest.scala:30)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at scala.util.Either.flatMap(Either.scala:352)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at kinesis.mock.api.DescribeStreamRequest.$anonfun$describeStream$1(DescribeStreamRequest.scala:27)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at cats.effect.IOFiber.runLoop(IOFiber.scala:337)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at cats.effect.IOFiber.asyncContinueSuccessfulR(IOFiber.scala:1330)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at cats.effect.IOFiber.run(IOFiber.scala:119)
2023-05-31 14:57:49 2023-05-31T12:57:49.200  INFO --- [functhread20] l.s.k.kinesis_mock_server  : at cats.effect.unsafe.WorkerThread.run(WorkerThread.scala:585)
```

There was no such error in the versions <2.1.0

### How are you starting LocalStack?

With a `docker run` command

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

```shell
docker run --rm --name ls-2-1 -v /Users/ayeganyan/workspace/sandbox/localstack2.1.0:/commands -v /var/run/docker.sock:/var/run/docker.sock -e LS_DEBUG=1 -p 4566:4566 localstack/localstack:2.1.0
```

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

```shell
awslocal cloudformation create-stack --stack-name kinesis-test --template-body file:///commands/kinesis-v1.yml --parameters ParameterKey=ResourceName,ParameterValue=kinesis-test
```
Here is the content of `kinesis-v1.yml`:
```yaml
Parameters:
  ResourceName:
    Type: String

Resources:
  KinesisStream:
    Type: AWS::Kinesis::Stream
    Properties:
      Name: { Ref: ResourceName }
```

### Environment

```markdown
- OS: Macos Ventura 13.4
- LocalStack: 2.1.0
```

**Comment by ayeganyan** [comment]
@thrau @dominikschubert 
Not sure if you are aware or not, but we have the some new error when running the same script, this was not happening in `2.1.0`:

```
An error occurred (InternalError) when calling the CreateStack operation (reached max retries: 4): exception while calling cloudformation.CreateStack: Traceback (most recent call last):
  File "/opt/code/localstack/localstack/aws/chain.py", line 90, in handle
    handler(self, self.context, response)
  File "/opt/code/localstack/localstack/aws/handlers/service.py", line 123, in __call__
    handler(chain, context, response)
  File "/opt/code/localstack/localstack/aws/handlers/service.py", line 93, in __call__
    skeleton_response = self.skeleton.invoke(context)
  File "/opt/code/localstack/localstack/aws/skeleton.py", line 154, in invoke
    return self.dispatch_request(context, instance)
  File "/opt/code/localstack/localstack/aws/skeleton.py", line 166, in dispatch_request
    result = handler(context, instance) or {}
  File "/opt/code/localstack/localstack/aws/skeleton.py", line 118, in __call__
    return self.fn(*args, **kwargs)
  File "/opt/code/localstack/localstack/aws/api/core.py", line 167, in operation_marker
    return fn(*args, **kwargs)
  File "/opt/code/localstack/localstack/services/cloudformation/provider.py", line 196, in create_stack
    resolved_parameters = param_resolver.resolve_parameters(
  File "/opt/code/localstack/localstack/services/cloudformation/engine/parameters.py", line 78, in resolve_parameters
    raise Exception(
Exception: Invalid. Needs to have either param specified or Default. (TODO)
```

Let me know if I should open another issue for this

**Comment by ayeganyan-atl** [comment]
This is the exact script which I am running with localstack:latest (sha 50a2e28a5fe4) 
and I am having the error:

kinesis-create.sh
```shell
#!/usr/bin/env bash

LS_IMAGE=docker.atl-paas.net/atlassian/nebulae/localstack:latest

CONTAINER=$(docker run -d --rm \
        -v "$(pwd)":/commands \
        -v /var/run/docker.sock:/var/run/docker.sock \
        -e DEBUG=1 \
        "$LS_IMAGE")

docker exec "$CONTAINER" awslocal cloudformation create-stack --stack-name kinesis-test --template-body file:///commands/kinesis-v1.yml --parameters ParameterKey=ResourceName,ParameterValue=kinesis-test
```
kinesis-v1.yml:
```yaml
Parameters:
  MicrosResourceName: # automatic
    Type: String

Resources:
  KinesisStream:
    Type: AWS::Kinesis::Stream
    Properties:
      Name: { Ref: MicrosResourceName }
```

Both files are in the same directory

**Comment by ayeganyan-atl** [comment]
@pinzon Don't you reproduce the initial error (`kinesis.mock.ResourceNotFoundException: Stream arn arn:aws:kinesis:us-east-1:000000000000:stream/dx-ref-cfn-kinesis1 not found`)?

Not sure yet, but seems that affecting lambdas that are subscribed to stream.

**Comment by pinzon** [comment]
Those initial logs are normal because some cfn resources will try to fetch the state even if the resource doesn't exist yet in the API.

You could try deploying again and then calling cfn describe-stacks and the stack status will appear as CREATE_COMPLETE, then you can list the streams and call a describe operation on the stream and the new stream should be there.

If you're having a problem with the kinesis-lambda integration that could be another issue.

**Comment by ayeganyan-atl** [comment]
In logs of our product I have the following logs:
```
2023-06-28 09:19:13 2023-06-28T07:19:13.089  INFO --- [   asgi_gw_0] localstack.utils.bootstrap : Execution of "_load_service_plugin" took 543.69ms
2023-06-28 09:19:13 2023-06-28T07:19:13.090  INFO --- [   asgi_gw_0] localstack.utils.bootstrap : Execution of "require" took 544.38ms
2023-06-28 09:19:13 2023-06-28T07:19:13.327 ERROR --- [   asgi_gw_0] l.aws.handlers.logging     : exception during call chain: Invalid. Needs to have either param specified or Default. (TODO)
2023-06-28 09:19:13 2023-06-28T07:19:13.328  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS cloudformation.CreateStack => 500 (InternalError)
2023-06-28 09:19:13 2023-06-28T07:19:13.985 ERROR --- [   asgi_gw_0] l.aws.handlers.logging     : exception during call chain: Invalid. Needs to have either param specified or Default. (TODO)
2023-06-28 09:19:13 2023-06-28T07:19:13.985  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS cloudformation.CreateStack => 500 (InternalError)
2023-06-28 09:19:14 2023-06-28T07:19:14.133 ERROR --- [   asgi_gw_0] l.aws.handlers.logging     : exception during call chain: Invalid. Needs to have either param specified or Default. (TODO)
2023-06-28 09:19:14 2023-06-28T07:19:14.134  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS cloudformation.CreateStack => 500 (InternalError)
2023-06-28 09:19:15 2023-06-28T07:19:15.367 ERROR --- [   asgi_gw_0] l.aws.handlers.logging     : exception during call chain: Invalid. Needs to have either param specified or Default. (TODO)
2023-06-28 09:19:15 2023-06-28T07:19:15.367  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS cloudformation.CreateStack => 500 (InternalError)
2023-06-28 09:19:19 2023-06-28T07:19:19.603 ERROR --- [   asgi_gw_0] l.aws.handlers.logging     : exception during call chain: Invalid. Needs to have either param specified or Default. (TODO)
2023-06-28 09:19:19 2023-06-28T07:19:19.603  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS cloudformation.CreateStack => 500 (InternalError)
2023-06-28 09:19:19 
2023-06-28 09:19:19 An error occurred (InternalError) when calling the CreateStack operation (reached max retries: 4): exception while calling cloudformation.CreateStack: Invalid. Needs to have either param specified or Default. (TODO)
2023-06-28 09:19:19 2023-06-28T07:19:19.869  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS cloudformation.CreateStack => 200
2023-06-28 09:19:19 {
2023-06-28 09:19:19     "StackId": "arn:aws:cloudformation:ap-south-1:000000000000:stack/kinesis-v1-dx-ref-cfn-kinesis/2215632f"
2023-06-28 09:19:19 }
```
So after some attempts stream is created, I wonder what does this mean.

Our tests still failing, investigating...

**Comment by ayeganyan-atl** [comment]
It turns out we had issue with some other resource parameters. Great work on validation of parameters!

**Comment by jim-meyer** [comment]
FWIW I ran into this same cryptic error and discovered that I was missing the `--parameter-overrides`. I was using:

`deploy -t template-db.yaml --no-confirm-changeset --no-fail-on-empty-changeset --capabilities CAPABILITY_NAMED_IAM --stack-name mystackname Param1=foo Param2=bar`

Once I added `--parameter-overrides` it started working:

`deploy -t template-db.yaml --no-confirm-changeset --no-fail-on-empty-changeset --capabilities CAPABILITY_NAMED_IAM --stack-name mystackname --parameter-overrides Param1=foo Param2=bar `

Seems undesirable for this to result in such a cryptic failure so deep into the code. Normally problems with command line args should result in near immediate failures.
